### PR TITLE
Add async overload for search API

### LINKED ISSUE #332: Add async/await support
**Description**
Since server-side Swift projects like Vapor now support async/await APIs, it would nice if meilisearch-swift also added this.

**Basic example**
As opposed to the completion parameter style currently used, an async/await call would look like so:

    let result = await index.updateDocuments(documents: [/** data **/], primaryKey: "id")
    switch result {
    case .success:
        // ....
    case .failure(let error):
        // ....
    }

**Proposed solution**
I have already created extensions in my local project to support some of the methods that I use. For example `updateDocuments()` is covered by this implementation:

    public func updateDocuments<T>(
      documents: [T],
      encoder: JSONEncoder? = nil,
      primaryKey: String? = nil) async -> Result<TaskInfo, Swift.Error> where T: Encodable {
          return await withCheckedContinuation { continuation in
              self.updateDocuments(documents: documents, encoder: encoder, primaryKey: primaryKey) { res in
                  continuation.resume(returning: res)
              }
          }
    }

Proposal is to add these bridging methods for _all_ methods on `Indexes` in perhaps an `Indexes+async.swift` extension. Additionally unit tests will need to be added to cover async calls.

### PR DESCRIPTION
**Pull request by Sherlouk**
# Pull Request

⚠️ I've marked this as draft as it current works to provide an example of how we could add async/await support to the library. I am more than happy to expand this to **all** currently supported APIs.

## Related issue

Relates to #332 (doesn't close as only adds support for one API)

## What does this PR do?
- Adds an async/await override to an existing search API providing users with more choice.

## PR checklist
Please check if your PR fulfills the following requirements:
- [x] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [x] Have you read the contributing guidelines?
- [x] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

File: Sources/MeiliSearch/Indexes.swift [hunk 1]
```diff
@@ -26,7 +26,7 @@ public struct Indexes {
   private let documents: Documents
 
   // Search methods
-  private let search: Search
+  fileprivate let search: Search
```
Comment (aronbudinszky) [inline Sources/MeiliSearch/Indexes.swift:29]: This change is actually not required - `private` will work just as fine.

**Review by aronbudinszky — APPROVED** [review]
Aside from the minor comment regarding `fileprivate` all looks good to me!

I have not actually tested it though, but assume it does work given that the unit test passes.

File: Sources/MeiliSearch/Indexes.swift (hunk 1 above)
Comment (Sherlouk) [reply Sources/MeiliSearch/Indexes.swift:29]: Let me double check this, I thought I had a compiler error when I tested (due to use of an extension) but will double check and revert if possible.

In respect of "I have not actually tested it though" - I'm using this in a production application and is working fine for me 🙂 Though I'll keep an eye on it!

**Comment by curquiza** [comment]
Thank you @aronbudinszky for reviewing and @Sherlouk for the PR

Can you confirm there is no breaking for the user? 😊 
Is this change worth adding some documentation to the README? Or is ti really "usage transparent"?

**Comment by Sherlouk** [comment]
This is not a breaking change, no.
I do think it's worth updating documentation with a quick example, but it is also obvious when using.

**Comment by Sherlouk** [comment]
I've updated documentation, tweaked implementation, and opened up the PR. This does only add support for one API (search), but if we're happy then I can either:

1. Merge this PR, and then open a new one with support for all other APIs
2. Keep this PR open, until I've added support for all other APIs
3. Merge this PR, but open support for each other API one PR at a time (as I work on them with other features).

Clem, thoughts?

**Comment by curquiza** [comment]
Let's go with #1, not with #3 indeed 😄

**Review by curquiza — APPROVED** [review]
bors merge

File: Sources/MeiliSearch/Indexes.swift (hunk 1 above)
Comment (aronbudinszky) [reply Sources/MeiliSearch/Indexes.swift]: Yeah that's fair, agreed. (If there is value in hiding these properties internally we could still use protocols for that - though not sure if it is necessary.)

Only final point that I'd make then is that I would not explicitly state `internal` - given that this is the default anyway and not used on other internal properties.

**Comment by aronbudinszky** [comment]
Btw retroactively confirming there is no breaking change. :)

File: Sources/MeiliSearch/Indexes.swift (hunk 1 above)
Comment (Sherlouk) [reply Sources/MeiliSearch/Indexes.swift]: Fair point, that's a personal preference but deviates from standard practice for the project. I'll amend in a follow up PR

**Pull request merged**